src/poetry/core/packages/utils/utils.py

```diff
@@ -138,20 +142,20 @@ def splitext(path: str) -> tuple[str, str]:
 
 def group_markers(
     markers: list[BaseMarker], or_: bool = False
-) -> list[tuple[str, str, str] | list[tuple[str, str, str]]]:
+) -> list[list[tuple[str, str, str]]]:
     from poetry.core.version.markers import MarkerUnion
     from poetry.core.version.markers import MultiMarker
     from poetry.core.version.markers import SingleMarker
 
-    groups = [[]]
+    groups: list[list[tuple[str, str, str]]] = [[]]
 
     for marker in markers:
         if or_:
             groups.append([])
 
         if isinstance(marker, (MultiMarker, MarkerUnion)):
             groups[-1].append(
-                group_markers(marker.markers, isinstance(marker, MarkerUnion))
+                group_markers(marker.markers, isinstance(marker, MarkerUnion))  # type: ignore[arg-type]
```

Comment:
I got stuck here (in combination with the changed return type), but maybe not worth thinking about it since `group_markers` is not required anymore with #347. (The output of it seems to be in a weird form for representing `or` anyway.)